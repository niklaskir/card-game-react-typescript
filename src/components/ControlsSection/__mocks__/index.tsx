import React, { FC, ReactNode } from 'react';

const ControlsSection: FC<ReactNode> = () => <div id="controls_section" />;

export default ControlsSection;
