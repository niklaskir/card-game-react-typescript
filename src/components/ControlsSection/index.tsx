import axios from 'axios';
import React from 'react';
import { isLoading as isLoadingState } from '../../recoil/states/isLoading';
import {
  userListSortType as userListSortTypeState,
  userSelected,
} from '../../recoil/states/userList';
import { useRecoilState, useRecoilValue, useSetRecoilState } from 'recoil';

const ControlsSection = () => {
  const selectedUser = useRecoilValue(userSelected);
  const setUserSort = useSetRecoilState(userListSortTypeState);
  const [isLoading, setIsLoading] = useRecoilState(isLoadingState);

  const handleAsc = () => {
    setUserSort('Asc');
  };

  const handleDesc = () => {
    setUserSort('Desc');
  };

  const handleSubmit = async () => {
    if (selectedUser === null) {
      alert('Please select user');
      return;
    }

    setIsLoading(true);
    await axios.post('https://reqres.in/api/users', selectedUser);
    setIsLoading(false);
  };

  return (
    <div className="col-span-1">
      <div className="rounded overflow-hidden shadow-lg">
        <div className="px-6 py-4 bg-slate-100">
          <div className="font-bold text-xl mb-2">Controls</div>
          <div className="grid grid-cols-2 gap-4">
            <div className="col-span-1">
              <button
                className="w-full bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4"
                onClick={handleAsc}
              >
                SORT ASC
              </button>
            </div>
            <div className="col-span-1">
              <button
                className="w-full bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4"
                onClick={handleDesc}
              >
                SORT DESC
              </button>
            </div>
            <div className="col-span-2">
              {isLoading ? (
                <button
                  disabled
                  className="button w-full bg-slate-500 text-white font-bold py-2 px-4 cursor-not-allowed"
                >
                  Loading...
                </button>
              ) : (
                <button
                  className="w-full bg-green-500 hover:bg-green-700 text-white font-bold py-2 px-4"
                  onClick={handleSubmit}
                >
                  SUBMIT
                </button>
              )}
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

export default ControlsSection;
