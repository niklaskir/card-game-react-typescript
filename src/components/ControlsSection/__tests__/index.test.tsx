import users from '../../../data.json';
import React from 'react';
import Recoil from 'recoil';
import renderer from 'react-test-renderer';
import ControlsSection from '../index';
import axios from 'axios';

jest.mock('recoil', () => ({
  atom: jest.fn(),
  selector: jest.fn(),
  useRecoilState: jest.fn(),
  useRecoilValue: jest.fn(),
  useSetRecoilState: jest.fn(),
}));

jest.mock('axios', () => ({
  post: jest.fn(),
}));

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useState: jest.fn(),
}));

describe('ControlsSection', () => {
  it('should render correctly when isLoading is false', () => {
    Recoil.useRecoilState = jest.fn().mockReturnValueOnce([false, jest.fn()]);

    const overviewSection = renderer.create(<ControlsSection />).toJSON();

    expect(overviewSection).toMatchSnapshot();
  });

  it('should render correctly when isLoading is true', () => {
    Recoil.useRecoilState = jest.fn().mockReturnValueOnce([true, jest.fn()]);

    const overviewSection = renderer.create(<ControlsSection />).toJSON();

    expect(overviewSection).toMatchSnapshot();
  });

  it('should call setUserSort with correct params when clicking asc button', () => {
    const setUserSortSpy = jest.fn();
    Recoil.useSetRecoilState = jest.fn().mockReturnValueOnce(setUserSortSpy);
    Recoil.useRecoilState = jest.fn().mockReturnValueOnce([false, jest.fn()]);

    const overviewSection = renderer.create(<ControlsSection />);

    overviewSection.root
      .findAllByProps({
        className:
          'w-full bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4',
      })[0]
      .props.onClick();

    expect(overviewSection.toJSON()).toMatchSnapshot();
    expect(setUserSortSpy).toHaveBeenCalledWith('Asc');
  });

  it('should call setUserSort with correct params when clicking desc button', () => {
    const setUserSortSpy = jest.fn();
    Recoil.useSetRecoilState = jest.fn().mockReturnValueOnce(setUserSortSpy);
    Recoil.useRecoilState = jest.fn().mockReturnValueOnce([false, jest.fn()]);

    const overviewSection = renderer.create(<ControlsSection />);

    overviewSection.root
      .findAllByProps({
        className:
          'w-full bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4',
      })[1]
      .props.onClick();

    expect(overviewSection.toJSON()).toMatchSnapshot();
    expect(setUserSortSpy).toHaveBeenCalledWith('Desc');
  });

  it('should call axios.post with correct params when clicking submit button with selectedUser', async () => {
    const selectedUser = users[0];
    const setIsLoadingSpy = jest.fn();
    Recoil.useRecoilState = jest
      .fn()
      .mockReturnValueOnce([false, setIsLoadingSpy]);
    Recoil.useRecoilValue = jest.fn().mockReturnValueOnce(selectedUser);

    const overviewSection = renderer.create(<ControlsSection />);

    await overviewSection.root
      .findAllByProps({
        className:
          'w-full bg-green-500 hover:bg-green-700 text-white font-bold py-2 px-4',
      })[0]
      .props.onClick();

    expect(setIsLoadingSpy).toHaveBeenCalledWith(true);
    expect(axios.post).toHaveBeenCalledWith(expect.anything(), selectedUser);
    expect(setIsLoadingSpy).toHaveBeenCalledWith(false);
  });

  it('should not call axios.post when clicking submit button with selectedUser is null', async () => {
    const jsdomAlert = window.alert;
    window.alert = jest.fn();
    const selectedUser = null;
    const setIsLoadingSpy = jest.fn();
    Recoil.useRecoilState = jest
      .fn()
      .mockReturnValueOnce([false, setIsLoadingSpy]);
    Recoil.useRecoilValue = jest.fn().mockReturnValueOnce(selectedUser);

    const overviewSection = renderer.create(<ControlsSection />);

    await overviewSection.root
      .findAllByProps({
        className:
          'w-full bg-green-500 hover:bg-green-700 text-white font-bold py-2 px-4',
      })[0]
      .props.onClick();

    expect(setIsLoadingSpy).not.toHaveBeenCalled();
    expect(axios.post).not.toHaveBeenCalled();
    expect(window.alert).toHaveBeenCalled();

    window.alert = jsdomAlert;
  });
});
