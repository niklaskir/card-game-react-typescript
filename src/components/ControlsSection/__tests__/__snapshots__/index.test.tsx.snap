// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlsSection should call setUserSort with correct params when clicking asc button 1`] = `
<div
  className="col-span-1"
>
  <div
    className="rounded overflow-hidden shadow-lg"
  >
    <div
      className="px-6 py-4 bg-slate-100"
    >
      <div
        className="font-bold text-xl mb-2"
      >
        Controls
      </div>
      <div
        className="grid grid-cols-2 gap-4"
      >
        <div
          className="col-span-1"
        >
          <button
            className="w-full bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4"
            onClick={[Function]}
          >
            SORT ASC
          </button>
        </div>
        <div
          className="col-span-1"
        >
          <button
            className="w-full bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4"
            onClick={[Function]}
          >
            SORT DESC
          </button>
        </div>
        <div
          className="col-span-2"
        >
          <button
            className="w-full bg-green-500 hover:bg-green-700 text-white font-bold py-2 px-4"
            onClick={[Function]}
          >
            SUBMIT
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ControlsSection should call setUserSort with correct params when clicking desc button 1`] = `
<div
  className="col-span-1"
>
  <div
    className="rounded overflow-hidden shadow-lg"
  >
    <div
      className="px-6 py-4 bg-slate-100"
    >
      <div
        className="font-bold text-xl mb-2"
      >
        Controls
      </div>
      <div
        className="grid grid-cols-2 gap-4"
      >
        <div
          className="col-span-1"
        >
          <button
            className="w-full bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4"
            onClick={[Function]}
          >
            SORT ASC
          </button>
        </div>
        <div
          className="col-span-1"
        >
          <button
            className="w-full bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4"
            onClick={[Function]}
          >
            SORT DESC
          </button>
        </div>
        <div
          className="col-span-2"
        >
          <button
            className="w-full bg-green-500 hover:bg-green-700 text-white font-bold py-2 px-4"
            onClick={[Function]}
          >
            SUBMIT
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ControlsSection should render correctly when isLoading is false 1`] = `
<div
  className="col-span-1"
>
  <div
    className="rounded overflow-hidden shadow-lg"
  >
    <div
      className="px-6 py-4 bg-slate-100"
    >
      <div
        className="font-bold text-xl mb-2"
      >
        Controls
      </div>
      <div
        className="grid grid-cols-2 gap-4"
      >
        <div
          className="col-span-1"
        >
          <button
            className="w-full bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4"
            onClick={[Function]}
          >
            SORT ASC
          </button>
        </div>
        <div
          className="col-span-1"
        >
          <button
            className="w-full bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4"
            onClick={[Function]}
          >
            SORT DESC
          </button>
        </div>
        <div
          className="col-span-2"
        >
          <button
            className="w-full bg-green-500 hover:bg-green-700 text-white font-bold py-2 px-4"
            onClick={[Function]}
          >
            SUBMIT
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ControlsSection should render correctly when isLoading is true 1`] = `
<div
  className="col-span-1"
>
  <div
    className="rounded overflow-hidden shadow-lg"
  >
    <div
      className="px-6 py-4 bg-slate-100"
    >
      <div
        className="font-bold text-xl mb-2"
      >
        Controls
      </div>
      <div
        className="grid grid-cols-2 gap-4"
      >
        <div
          className="col-span-1"
        >
          <button
            className="w-full bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4"
            onClick={[Function]}
          >
            SORT ASC
          </button>
        </div>
        <div
          className="col-span-1"
        >
          <button
            className="w-full bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4"
            onClick={[Function]}
          >
            SORT DESC
          </button>
        </div>
        <div
          className="col-span-2"
        >
          <button
            className="button w-full bg-slate-500 text-white font-bold py-2 px-4 cursor-not-allowed"
            disabled={true}
          >
            Loading...
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
